AP_SteerController: Steer limit on speed

EShamaev opened the pull request:
AP_SteerController: added derating of steer angle depending on speed

When using steering controller on my plane (big - 30 kg, 3.6m wingspan) at high speed sometimes controller oversteers that leads to rolling on the wing.
Derating of steering was added.
Below certain speed - no derating, then as speed raises - maximum throw is lowered until defined minimum throw is reached.

muramura commented on libraries/APM_Control/AP_SteerController.cpp:
```diff
@@ -156,11 +184,21 @@ int32_t AP_SteerController::get_steering_out_rate(float desired_rate)
+    float derate_constraint = 4500;
+
+    // Calculate required constrain based on speed
+    if(speed > _deratespeed) {
```
this if and other if statement , in white space.

EShamaev replied:
Done.

muramura commented on libraries/APM_Control/AP_SteerController.cpp:
```diff
@@ -83,6 +83,34 @@ const AP_Param::GroupInfo AP_SteerController::var_info[] = {
+    // @Increment: 0.1
+    // @Units: Centidegrees
+    // @User: Advanced
+    AP_GROUPINFO("DRTMIN", 10, AP_SteerController, _deratefactor,        4500),
```
DRTFCT and DRTMIN, variable name equal ok?

EShamaev replied:
That is typo during update...

OXINARF commented:
@EShamaev Please read ardupilot.org/dev/docs/submitting-patches-back-to-master.html  You are not following our rules, namely a clean history (no merge commits and squashing fixes) and proper commit messages.

EShamaev commented:
Should I delete this branch and make a new one (without merging the official ArduPilot/ardupilot)?

OXINARF commented:
No need to delete the branch, just force push to it. You shouldn't merge, you should rebase your work on master.

EShamaev commented:
rebased, force pushed

OXINARF commented:
You haven't rebased on master and deleted the merge commits. Search for interative rebase to learn how to use it. If you need help come to Gitter.

EShamaev commented:
Like this?

OXINARF commented:
Yes, that's better. You still haven't the proper commit message, but that's fine for now, just remember it next time.

EShamaev commented:
Alright, understood about github, fill follow that rules.
But still can't understand what's wrong with message? I corrected and added Subsystem and brief and long description...

OXINARF commented:
Oh I see what you did, you changed the GitHub pull request title. I'm referring to the commits (in this case you only have one). Your commit message is just:

`Added derating of steering wheel`

but should be:

`APM_Control: added derating of steering wheel` or `AP_SteerController: added derating of steering wheel`

EShamaev commented:
Understood! Ok, wil follow!

EShamaev commented:
I based this on my experience.
Also next major update I will show is the taxi waypoints and as it is already done and now under tests, this problem showed there.
We need to know the speed after which we need to derate. Also the amount of derating. And some must be left so we have at least minor control on high speed.
I looked over all parameters, but not much is about steering.

khancyr commented on libraries/APM_Control/AP_SteerController.cpp:
```diff
@@ -156,11 +184,21 @@ int32_t AP_SteerController::get_steering_out_rate(float desired_rate)
+    float derate_constraint = 4500;
+
+    // Calculate required constrain based on speed
+    if (speed > _deratespeed) {
```
at _deratespeed = 0 , derating is always applied , this is in contradiction with "If 0 then no derating is used."

EShamaev replied:
Noted, thank you.

khancyr commented:
@OXINARF we should renable autotest for Rover as this PR will impact Rover too !

EShamaev commented:
Default values should prevent this.

khancyr commented:
@EShamaev yep now it is good!

OXINARF commented:
> @OXINARF we should renable autotest for Rover as this PR will impact Rover too !

Autotest in CI was never enabled for Rover. Feel free to open a PR to add it.

gmorph commented on libraries/APM_Control/AP_SteerController.cpp:
```diff
@@ -83,6 +83,34 @@ const AP_Param::GroupInfo AP_SteerController::var_info[] = {
+
+    // @Param: DRTMIN
+    // @DisplayName: Minimum angle of wheel
+    // @Description: The angle that limits smalles angle of steering wheel at maximum speed. Even if it should derate below, it will stop derating at this angle.
```
smalles is spelt wrong I believe

gmorph commented:
These parameters will be visible to Rover users as well.  It would be good to have the descriptions for the parameters suitable for Rover's as well as Plane's - I'm happy to help with the wording.  My understanding of the code is that once the vehicle reaches a certain speed - DRTSPD - turn angle is reduced - DRTFCT - is that correct?  I don't quite follow what DTRMIN is for - could you clarify please?
Thanks, Grant.

EShamaev commented:
DRTSPD - speed after which angle gets reduced
DRTFCT - amount in degrees to reduce for each additional m/s
DTRMIN - minimum angle that is left

![image](https://cloud.githubusercontent.com/assets/7996972/21291759/d1097766-c4f4-11e6-856a-2a5bf14aa36e.png)

EShamaev commented:
Rebased and pushed to check on rover.

gmorph commented:
I see - its the minimum angle the steering can be reduced to.  Love that picture - it should go onto the wiki as it makes it very clear what is going on.
LGTM
Thanks, Grant.
